### PR TITLE
[AST] Preserve sugar for generic typealiases.

### PR DESCRIPTION
**Pull request by dcci**
<rdar://problem/43110802>

### CODE REVIEW

**Comment by dcci** [comment]
@swift-ci test and merge

**Comment by jrose-apple** [comment]
Isn't this not correct because we've lost the parent type? Or am I misremembering? @slavapestov

**Comment by dcci** [comment]
@swift-ci please clean test

**Comment by dcci** [comment]
@swift-ci please clean test macos

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -3022,14 +3022,13 @@ void TypeAliasDecl::setUnderlyingType(Type underlying) {
...
-      setInterfaceType(MetatypeType::get(sugaredType, ctx));
-    } else {
-      setInterfaceType(MetatypeType::get(underlying, ctx));
-    }
+    auto *genericSig = getGenericSignature();
+    auto subs = SubstitutionMap::get(
+        genericSig, [&](SubstitutableType *type) -> Type { return type; },
+        MakeAbstractConformanceForGenericType());
+
+    auto sugaredType = NameAliasType::get(this, Type(), subs, underlying);
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: The parent type should probably be the declared interface type of the parent DC, if the parent is a type context

**Pull request closed without merging**